core/rio/turtle/src/main/java/org/eclipse/rdf4j/rio/turtle/TurtleWriter.java

```diff
@@ -701,4 +735,109 @@ private void wrapLine(boolean space) throws IOException {
+					processSubject(contextData, nextSubject.get(), processedSubjects);
+					nextSubject = nextSubject(contextData, processedSubjects);
+				}
+			}
+		} else {
+			// group by subject
+
+			Set<Resource> processedSubjects = new HashSet<>();
+			Optional<Resource> nextSubject = nextSubject(bufferedStatements, processedSubjects);
+			while (nextSubject.isPresent()) {
+				processSubject(bufferedStatements, nextSubject.get(), processedSubjects);
+				nextSubject = nextSubject(bufferedStatements, processedSubjects);
+			}
+		}
+		bufferedStatements.clear();
+	}
+
+	private Optional<Resource> nextSubject(Model contextData, Set<Resource> processedSubjects) {
+		for (Resource subject : contextData.subjects()) {
+			if (processedSubjects.contains(subject)) {
+				continue;
+			}
+			if (subject instanceof BNode && inlineBNodes) {
+				Set<Resource> otherSubjects = contextData.filter(null, null, subject).subjects();
+				if (otherSubjects.stream().anyMatch(s -> !processedSubjects.contains(s))) {
+					continue;
+				}
+			}
+			return Optional.of(subject);
+		}
+		return Optional.empty();
+	}
+
+	private void processSubject(Model contextData, Resource subject, Set<Resource> processedSubjects) {
+		if (processedSubjects.contains(subject)) {
+			return;
+		}
+
+		Set<IRI> processedPredicates = new HashSet<>();
+
+		// give rdf:type preference over other predicates.
+		for (Statement typeStatement : contextData.getStatements(subject, RDF.TYPE, null)) {
+			boolean canInlineObject = inlineValue(contextData, typeStatement.getObject());
+			handleStatementInternal(typeStatement, false,
+					inlineValue(contextData, typeStatement.getSubject()),
+					canInlineObject);
+
+			if (canInlineObject && typeStatement.getObject() instanceof BNode) {
+				processSubject(contextData, (BNode) typeStatement.getObject(), processedSubjects);
+			}
+		}
+
+		processedPredicates.add(RDF.TYPE);
+
+		// retrieve other statement from this context with the same
+		// subject, and output them grouped by predicate
+		for (Statement subjectStatement : contextData.getStatements(subject, null, null)) {
+			IRI predicate = subjectStatement.getPredicate();
+			if (!processedPredicates.contains(predicate)) {
+				for (Statement toWrite : contextData.getStatements(subject, predicate, null)) {
+					boolean canInlineObject = inlineValue(contextData, toWrite.getObject());
+					handleStatementInternal(toWrite, false, inlineValue(contextData, toWrite.getSubject()),
+							canInlineObject);
+					if (canInlineObject && toWrite.getObject() instanceof BNode) {
+						processSubject(contextData, (BNode) toWrite.getObject(), processedSubjects);
+					}
+				}
+				processedPredicates.add(predicate);
+			}
+
+		}
+
+		processedSubjects.add(subject);
+	}
+
+	private boolean inlineValue(Model contextData, Value v) {
+		if (!inlineBNodes) {
+			return false;
+		}
+		if (v instanceof BNode) {
+			return (contextData.filter(null, null, v).size() <= 1);
+		}
+		return true;
+	}
```

Comment:
Maybe this should be `canInlineValue` to indicate that it checks if inlining is allowed.

Reply:
It's a nice and simple approach though. I like it a lot.